### ISSUE TITLE
Column searching does not combine multiple columns properly

### ISSUE THREAD

**Issue by JimHough** [aldeed/meteor-tabular#79]
Hi,
I'm trying to implement per-column searching as in [this](http://datatables.net/examples/api/multi_filter.html) example.
As the data is set to be server side tabular ignores the per column search data.
Can I use the selector to accomplish the same effect?

Thanks

**Comment by mrn3** [comment]
I am trying to do this too - anyone have some example code in Meteor?

**Comment by petrometro** [comment]
+1

**Comment by kirilan** [comment]
there is a plugin for datatables for filtering that can do something like this but it won't work for this module because of the way it handles data on the server side

**Comment by mrn3** [comment]
That is a bummer. I guess I will just do custom filtering using template helpers.  Thanks.

**Comment by jasonxeno** [comment]
Has anyone found any solutions for accomplishing this or simulating this?

**Comment by JimHough** [comment]
I think the key is to re-actively change the selector based on the values on the per column text boxes. I'm still working on this but other tasks have taken priority for the moment.

**Comment by petrometro** [comment]
A bit of a hack but here's how I've got this working for the moment... followed the example [here](https://datatables.net/examples/api/multi_filter.html). Basically just using the fact that the selector attribute is reactive...

```javascript
Template.TableContainer.created = function(){
    this.currentSelector = new ReactiveVar({ /* default client selector here */ });
}

Template.TableContainer.rendered = function(){
    var self = this,
          $table = $('.myTable'),
          $input = $table.find('thead th'); // assuming one column, follow the example for multiple columns

    $input.on('keyup', function(){
        var sel = $.clone(self.currentSelector.get());
        delete sel.my_column_name;
        sel.my_column_name = this.value;
        self.currentSelector.set(sel);
    });
}

Template.TableContainer.helpers({
    currentSelector : function(){
        return Template.instance().currentSelector.get();
    }
});
```
```html
<template name='TableContainer'>
    {{>tabular selector=currentSelector ... .... }}
</template>```

**Comment by miri-am** [comment]
@petrometro. Thanks for your example. However cannot get the filtering to work. What do I need to put into ```/* default client selector here */```? Trying to make the first column filterable... Thanks for any hints.

 I have the following: 
```javascript
//tables.js in common code
TabularTables = {};

Meteor.isClient && Template.registerHelper('TabularTables', TabularTables);

TabularTables.Audit = new Tabular.Table({
  name: "AuditList",
  collection: Audit,
  columns: [
    {data: "type", title: "Typ", class: "my_col"},
    {data: "timestamp", title: "Zeit"},
    {data: "collection", title: "Collection"}
  ]
});
```

```javascript
// client.js
Template.TableContainer.created = function(){
    this.currentSelector = new ReactiveVar({ /* default client selector here */ });
}

Template.TableContainer.rendered = function(){
    var self = this,
          $table = $('#tabuTable'),
          $input = $('<input name="type" type="text" placeholder="Filter"/>');

    $table.last('thead th.my_col').append($input); // column class can be set in TabularTable definition

    $input.on('keyup', function(){
        var sel = self.currentSelector.get();
        if(this.value) {
            sel.my_col = { $regex : this.value, $options : 'i' };
        } else {
            delete sel.my_col;
        }
        self.currentSelector.set(sel);
    });
}

Template.TableContainer.helpers({
    currentSelector : function(){
        return Template.instance().currentSelector.get();
    }
});
```

```html
<template name='TableContainer'>
      {{> tabular table=TabularTables.Audit selector=currentSelector class="table table-striped table-hover" id="tabuTable"}}
</template>
```

**Comment by petrometro** [comment]
hello @miri-am ok one thing to try would be to stop the event propagation on the keyup listener (see my code above)... maybe the event is getting swallowed by the column sort listener...

the default selector could just be {}. But if you *always* want to limit your documents to e.g. timestamp > 1420070400, you could just put { timestamp : { $gt : 1420070400 }, and then if you type 'foo' into the column search input your mongo search would be
```
{ timestamp : { $gt : 1420070400 }, type : { regex : 'foo', $options : 'i' }}
```

**Comment by miri-am** [comment]
ah, thanks, I found the problem. ```sel.my_col```was not set correctly. Now it is working fine. Thanks.

**Comment by AdamBrodzinski** [comment]
@miri-am , @petrometro , @jasonxeno  Not sure if this would help in your use cases but I have a fork/PR that allows you to modify the search selector with a callback before it's published (I needed to modify on the server). If your search string has something like "state: CA" then you can parse the string and narrow the selector to just the "state" column.

```js
TabularTables.Posts = new Tabular.Table({
  // other properties...

  modifySelector: function(selector) {
    // modify it here ...
    return selector;
  }

});
```

https://github.com/AdamBrodzinski/meteor-tabular/tree/change-selector-opt#modifying-the-selector

**Comment by miri-am** [comment]
@AdamBrodzinski. Thanks, looks good. Isnt this an answer to my other question? #114

**Comment by AdamBrodzinski** [comment]
@miri-am I'm not sure if it would work as-is for #114 with helpers because the callback runs serverside only... though I haven't needed helpers yet so it may work.

However you may be able to split the search string "John Doe" into two strings and add them as a query in the `$or` array... that would work. However doing that blindly without knowing what the intention is could return bad data. However if they searched "fullname: John Doe" it would be trivial to split them up that way.

**Comment by riiiiizzzzzohmmmmm** [comment]
This is because **tabular** only parses the global (global in the context of **DataTables**, that is) `{ search }` object, whereas **DataTables** attaches a column-specific `{ search }` object to each column under `{ ajax: { data: function(data) { data.columns.search ... } } }`.

To support **DataTables** column-specific search, I modified `/client/tabular.js`. In the `.rendered()` callback, I changed:

``` javascript
// Update pubSelector
var pubSelector = getPubSelector(
  template.tabular.selector,
  (data.search && data.search.value) || null,
  template.tabular.searchFields,
  template.tabular.searchCaseInsensitive
);
```

to 

``` javascript
// Update pubSelector
var pubSelector = getPubSelector(
  template.tabular.selector,
  (data.search && data.search.value) || null,
  template.tabular.searchFields,
  template.tabular.searchCaseInsensitive,
  data.columns || null
);
```

And I modified `/client/pubSelector.js` like so:

``` javascript
getPubSelector = function getPubSelector(selector, searchString, searchFields, searchCaseInsensitive, columns) {

  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  // if search was invoked via .columns().search(), build a query off that
  // https://datatables.net/reference/api/columns().search()
  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  var searchColumns = _.filter(columns, function(column) {
    return column.search && column.search.value !== '';
  });

  // required args
  if ((!searchString && searchColumns.length === 0) || !searchFields || searchFields.length === 0) {
    return selector;
  }

  // See if we can resolve the search string to a number,
  // in which case we use an extra query because $regex
  // matches string fields only.
  var searches = [];

  // normalize search fields array to mirror the structure 
  // as passed by the datatables ajax.data function
  searchFields = _.map(searchFields, function(field) {
    return { 
      data: field, 
      search: { 
        value: searchString 
      } 
    }; 
  });

  var searchTerms = _.isEmpty(searchColumns) ? searchFields : searchColumns;

  _.each(searchTerms, function(field) {
    var m1 = {}, m2 = {};

    var numSearchString = Number(field.search.value);

    // String search
    m1[field.data] = { $regex: field.search.value };

    // DataTables searches are case insensitive by default
    if (searchCaseInsensitive !== false) {
      m1[field.data].$options = "i";
    }

    searches.push(m1);

    // Number search
    if (!isNaN(numSearchString)) {
      m2[field.data] = numSearchString;
      searches.push(m2);
    }
  });

  return _.extend({}, selector, { $or: searches });
};
```

Basically, I just add `columns` as a 5th argument, and then if any of the columns has a non-empty search value (which is the case when using `table.columns().search()`), the query/selector object is built from that instead of the global search object. 

Otherwise, the global search object is used as usual. The query/selector still gets assembled in the exact same way. I shifted a few steps of the function around (like checking for a numeric search value).

I'm going to submit a PR, but figured I'd post this here for now.

**Comment by brylie** [comment]
I really like the interface of the Django Suit administration table view:

![Django Suit administration table view](
https://raw.github.com/darklow/django-suit/develop/docs/_static/img/django-suit.png)

**Comment by KyleKing** [comment]
@petrometro thanks for posting your code! I wanted to contribute a small addition that prevents the column from resorting when the user clicks the input box:
```
$input.on('click', function(e) {
  return e.stopPropagation();
});
```

**Comment by aldeed** [comment]
@riiiiizzzzzohmmmmm's fix is in the latest release but I haven't tested it yet. Can anyone confirm that column search works properly now?

**Comment by NicoKnoll** [comment]
@aldeed: Not working because I can't access .column() from client side...

**Comment by aldeed** [comment]
@NicoKnoll You can get DT instance by calling `$(theTableElement).DataTable()` so it would be `$(theTableElement).DataTable().column(0).search('searchTerm').draw()`

**Comment by voidale** [comment]
@aldeed Hey aldeed, I'm trying to use this but it seems that it's not combined filter it's searching just for one last row if you take a look at http://datatables.net/examples/api/multi_filter.html it's a merged filter you can filter office: new york + age: 40 

can we add this functionality? Or it should work already?

**Comment by voidale** [comment]
@aldeed Hey aldeed, I'm trying to use this but it seems that it's not combined filter it's searching just for one column if you take a look at http://datatables.net/examples/api/multi_filter.html it's a merged filter you can filter office: new york + age: 40

can we add this functionality? Or it should work already?

**Comment by voidale** [comment]
@aldeed any update on my last comment can we improve the filters? Thank you!

**Comment by timstew** [comment]
I would love to be able to get this working with $and as well. Has anyone figured it out? @voidale your pad was very helpful for me to get close (I like your form injection system!), but I guess we'll need to dig into the internals of this to try to change... it would be a great thing to expose as an option during instatiation... +1 multi column search with $and / $or - Might this kind of thing be on your roadmap @aldeed ?

**Comment by aldeed** [comment]
@voidale @timstew I will accept a PR on this. The code to look at is here: https://github.com/aldeed/meteor-tabular/blob/master/client/pubSelector.js#L9-L71

**Comment by gheller1** [comment]
This sounds like great functionality that I would love to be able to add to my project.  Can someone post a simple example as to how to implement per column searching with multiple columns with Tabular?  Thanks!

**Comment by akiarostami** [comment]
This is not a good solution for user-facing pages, but it's probably good for admins.  Using a small piece of code, I added a little extra feature to search: you can use ":" to define the field you want to search. For example, to search the fiend username only, you can use "username:admin".  I did this by removing all the other fields from the search selector when I see a ":" except from the the matching field.

Multiple word search works fine with this, assuming you are using OR for different search words.  You can simple change the code to use AND if that's your default search setting.

Wherever you define you  TabularTable, add this:

```
changeSelector: function(selector, userId) {
    var searchItems;
    var searchPhrase;
    var searchField;

    if (selector["$or"]) {
        searchItems = _.map(selector["$or"], function(item) {
            searchPhrase = _.values(item)[0]["$regex"];
            searchField = searchPhrase.substring(0, searchPhrase.indexOf(":"));
            if (searchField) {
                if (searchField === _.keys(item)[0])  // searchFiled matches the key (field name)
                    item[searchField]["$regex"] = searchPhrase.substring(searchPhrase.indexOf(":")+1);
                else
                    item = null;
            };

            return item;
        });

        searchItems = _.reject(searchItems, function(item) { return item === null })

        if (searchItems.length > 0)
            selector["$or"] = searchItems;
        else
            selector = null;
    }

    return selector;
};
```

**Comment by aldeed** [comment]
I believe this was fixed at some point.
